activestorage/README.md

```diff
@@ -106,6 +106,38 @@ Variation of image attachment:
 <%= image_tag user.avatar.variant(resize_to_limit: [100, 100]) %>
 ```
 
+## Delivery methods
+You can change the delivery method to best suit the needs of your application.
+
+### Redirect (default)
+Requests for files will redirect to a temporary service URL.
+
+### Proxy
+Files are proxied through the application server so they appear as though they're being served from your server. Useful for serving images from a CDN or HTML caching.
+
+### Changing delivery method
+globally
+```ruby
+config.active_storage.default_delivery_method = :proxy
+```
+model
+```ruby
+has_one_attached :avatar, delivery_method: :proxy
+```
+view
+```ruby
+user.avatar.variant(resize: "100x100").deliver(:proxy)
+```
+
+### Changing the host of proxied urls
+When using the proxy option to deliver assets you can set the host. This is useful if your CDN is setup to operate on a different domain.
+
+```ruby
+config.active_storage.delivery_methods = {
+  proxy: ActiveStorage::DeliveryMethod::Proxy.new(host: 'cdn.domain.com')
+}
```

Comment:
Do we have a lot of other ideas for delivery methods? I think going generic on this interface is complicating the feature more than its helping it. For the specific specialization of the host, that could just be a concrete feature of a proxy setup. You already have a `proxy_urls_expire_in` option.

I don't think this is carrying its weight.

Reply:
If you just go with proxy_url, in addition to service_url, I think a lot of this complicated setup goes away.

Reply:
I think you could actually junk a lot of the changes if you just committed to two paths: redirection and proxy. Then the representation controller could redirect to either Redirection or Proxy controller, based on a top-level setting, and you wouldn't have to be passing this URL stuff around everywhere.

Reply:
I'd like to see a concrete implementation of this idea that we find compelling enough to keep. Because the added complexity of this adapterized implementation is significant. Don't want to pay that price unless the payoff is for sure there.

Reply:
Here is a quick sketch of how [imgproxy](https://imgproxy.net) delivery adapter might look like:

```ruby
# production.rb

# using the currently provided API
config.active_storage.delivery_methods = {
  images_proxy: ActiveStorage::DeliveryMethod::Imgproxy.new(imgproxy_url: 'my.imgproxy.net', **credentials)
}

# development.rb / test.rb

# here we can use the built-in proxy adapter—that's where adapterization becomes especially useful;
# the same we do with storage backends
config.active_storage.delivery_methods = {
  images_proxy: ActiveStorage::DeliveryMethod::Proxy.new
}


# imgproxy delivery method
module ActiveStorage
  module DeliveryMethod
    class Imgproxy < Proxy
      def initialize(url:, **options)
        @imgproxy = Imgproxy::Builder.new(url, options)
      end

     def representation_url(representation, url_options: {})
       @imgproxy.url_for(
         # use rails endpoint as the source for image proxy
         super,
         # transform Active Storage transformations to proxy params
         transformation_to_options(representation.transformation)
      )
    end

    def blob_url(*)
      # original are passed as is
      super
    end
  end
end
```

The biggest benefit of this approach is interoperability (maybe, not complete, but at least API-wise). Users can continue using familiar Active Storage APIs, third-party developers don't need to monkey-patch Rails (like [they do](https://github.com/imgproxy/imgproxy.rb) right now).

Reply:
And there are many more such specialized proxies out there: [Cloudinary](https://cloudinary.com/solutions/image_management), [Imgix](https://www.imgix.com) to name a few.

If we add this API to Rails, using any of them will be as easy as using Active Storage itself. For both library developers (adding adapter class is easier and safer then monkey-patching) and end-users.

Reply:
> the added complexity of this adapterized implementation is significant

@dhh a lot of the complexity comes from adding the model level API `has_one_attached :avatar, delivery_method: :proxy`, not the adapterized approach. I'll remove the model API and apply your other suggestions and see if we can reach an agreeable solution.